044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,198 @@
+* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
+(i.e. not in pending state) before rolling the pod or waiting for it to become ready.
+
+The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
+The brokers do not move to RUNNING until a quorum has been formed.
+For example in combined mode during normal startup the following would happen:
+
+* All the pods are started at the same time
+* The controller processes in each pod form a quorum
+* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to RUNNING and are marked as ready
+* All pods are now ready
+
+However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:
+
+* The first pod is scheduled
+* The controller process starts up but cannot form a quorum because the other controllers are missing
+* The broker process starts up but since there is no controller quorum it does not move to RUNNING, it stays in STARTING. This means the pod is not marked as ready
+* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff
+
+## Proposed order of tasks to complete this proposal
+
+The KRaft liveness and readiness probes can be implemented in 3 different phases.
+Phases 1 and 2 can be implemented solely based on this proposal.
+Phase 3 can be implemented only after additional improvements of the `KafkaRoller` which will be covered by a separate proposal.
+
+### Phase 1 - Strimzi only supports combined mode
+* Combined mode pods are marked as both alive and ready when there is a Java process running. 
+This is an improvement on the current state where the pod is marked as alive and ready as soon as the container starts up, regardless of whether the Java process started.
+* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
+This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.
+
+### Phase 2 - Strimzi adds support for broker only and controller only modes
+* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
+* Broker only pods are marked as both alive and ready if listening on port 9091.
+* Combined mode pods are marked as both alive and ready when there is a Java process running (no change from phase 1).
+
+### Phase 3 - KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods
+* Broker only readiness probe fully implemented to match this proposal.
+* Combined mode readiness probe fully implemented to match this proposal.
```

Comment:
Just formatting improvements.

```suggestion
### Phase 1 - Strimzi only supports combined mode

* Combined mode pods are marked as both alive and ready when there is a Java process running. 
  This is an improvement on the current state where the pod is marked as alive and ready as soon as the container starts up, regardless of whether the Java process started.
* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
  This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.

### Phase 2 - Strimzi adds support for broker only and controller only modes

* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
* Broker only pods are marked as both alive and ready if listening on port 9091.
* Combined mode pods are marked as both alive and ready when there is a Java process running (no change from phase 1).

### Phase 3 - KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods

* Broker only readiness probe fully implemented to match this proposal.
* Combined mode readiness probe fully implemented to match this proposal.
```